tests/spec/corpus.rs

```diff
@@ -0,0 +1,226 @@
+use bson::Bson;
+use pretty_assertions::assert_eq;
+use serde_derive::Deserialize;
+
+use super::run_spec_test;
+
+#[derive(Debug, Deserialize)]
+struct TestFile {
+    description: String,
+    bson_type: String,
+    test_key: Option<String>,
+
+    #[serde(default)]
+    valid: Vec<Valid>,
+
+    #[serde(rename = "decodeErrors")]
+    #[serde(default)]
+    decode_errors: Vec<DecodeError>,
+
+    #[serde(rename = "parseErrors")]
+    #[serde(default)]
+    parse_errors: Vec<ParseError>,
+
+    deprecated: Option<bool>,
+}
+
+#[derive(Debug, Deserialize)]
+struct Valid {
+    description: String,
+    canonical_bson: String,
+    canonical_extjson: String,
+    relaxed_extjson: Option<String>,
+    degenerate_bson: Option<String>,
+    degenerate_extjson: Option<String>,
+    converted_bson: Option<String>,
+    converted_extjson: Option<String>,
+    lossy: Option<bool>,
+}
+
+#[derive(Debug, Deserialize)]
+struct DecodeError {
+    description: String,
+    bson: String,
+}
+
+#[derive(Debug, Deserialize)]
+struct ParseError {
+    description: String,
+    string: String,
+}
+
+fn run_test(test: TestFile) {
+    if test.bson_type == "0x13" && !cfg!(feature = "decimal128") {
+        return;
+    }
+
+    for valid in test.valid {
+        let description = format!("{}: {}", test.description, valid.description);
+
+        let bson_to_native_cb = bson::decode_document(
+            &mut hex::decode(&valid.canonical_bson)
+                .expect(&description)
+                .as_slice(),
+        )
+        .expect(&description);
+
+        let mut native_to_bson_bson_to_native_cv = Vec::new();
```

Comment:
The long names like this map directly to the descriptions in the corpus test README; they're annoying to read in a vaccuum, but they make it much easier to verify that we're in fact testing what the corpus tests are asking us to.